api/s2n.h

```diff
@@ -596,6 +596,210 @@ extern int s2n_cert_get_utf8_string_from_extension_data_length(const uint8_t *ex
 S2N_API 
 extern int s2n_cert_get_utf8_string_from_extension_data(const uint8_t *extension_data, uint32_t extension_len, uint8_t *out_data, uint32_t *out_len);
 
+/* PSK Hash Algorithm - RFC 8446 Section-2.2 */
+/* https://datatracker.ietf.org/doc/html/rfc8446#section-2.2*/
+typedef enum {
+    S2N_PSK_HMAC_SHA256,
+    S2N_PSK_HMAC_SHA384,
+} s2n_psk_hmac;
+
+struct s2n_psk;
+
+/**
+ * Creates a new s2n external PSK object with `S2N_PSK_HMAC_SHA256` as the default PSK hash algorithm.
+ * 
+ * # Safety 
+ *
+ * Use `s2n_psk_free` to free the memory allocated to the s2n external PSK object created by this API. 
+ *
+ * @return struct s2n_psk* Returns a pointer to the newly created external PSK object.
+ */
+S2N_API
+struct s2n_psk* s2n_external_psk_new();
+
+/**
+ * Frees the memory associated with the `s2n_psk` object.
+ *
+ * @param psk Pointer to the PSK object to be freed.
+ */
+S2N_API 
+int s2n_psk_free(struct s2n_psk **psk);
+
+/**
+ * Sets the PSK identity for a given PSK object. 
+ * 
+ * @param psk A pointer to PSK object to be updated with the PSK secret.
+ * @param identity The PSK identity being set.
+ * @param identity_size The length of the PSK identity being set.
+ */
+S2N_API 
+int s2n_psk_set_identity(struct s2n_psk *psk, const uint8_t *identity, uint16_t identity_size);
+
+/**
+ * Sets the PSK secret for a given PSK object. 
+ * 
+ * @param psk A pointer to PSK object to be updated with the PSK secret.
+ * @param secret The PSK secret in hex-encoded format.
+ * @param secret_size The length of the PSK secret being set.
+ */
+S2N_API 
+int s2n_psk_set_secret(struct s2n_psk *psk, const uint8_t *secret, uint16_t secret_size);
+
+/**
+ * Sets the PSK HMAC algorithm for a given PSK object. 
+ * 
+ * @param psk A pointer to the PSK object to be updated with the PSK HMAC algorithm.
+ * @param hmac The PSK HMAC algorithm being set. 
+ * The supported PSK HMAC algorithms are listed in the enum `s2n_psk_hmac` above. 
+ */
+S2N_API 
+int s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac hmac);
+
+/**
+ * Appends a PSK to the s2n connection object. If a duplicate PSK identity is found, 
+ * S2N_ERR_DUPLICATE_PSK_IDENTITIES error is thrown.
+ * 
+ * @param conn A pointer to the s2n_connection object to be updated with the PSK.
+ * @param psk A pointer to the PSK object to be appended.
+ */
+S2N_API 
+int s2n_connection_append_psk(struct s2n_connection *conn, struct s2n_psk *psk);
+
+/* s2n PSK supported modes for TLS1.3 */ 
```

Comment:
I think Madeleine had a good point in her documentation: we should assume these are for >=TLS1.3 / >TLS1.2.